plugins: Add ccc (new PR)

janurskremer opened the pull request.

MattSturgeon commented:
> > The option `plugins.ccc.default_color' does not exist.
> What does this error mean?

The test is malformed. Probably an issue with my suggestion. I suspect you need to nest your non-enable options in a `settings` block;

```
plugins.ccc.settings.*
```

MattSturgeon commented:
> I think I need to create a new pull request. At least I don't see how I could append to this one

You would've been able to _either_ push up more commits implementing the changes _or_ used one of many methods of amending history followed by a `--force` push.

History can be edited several ways, such as 

- `git commit --amend` to update the latest commit, `git reset --soft <some ref>`
-  `git reset --mixed <some ref>`, followed by making a new commit
- `git rebase --interactive <some ref>` to do more advanced stuff

You can also use lazygit's interface to abstract a lot of that.

I'd highly recommend learning some of gits fundamentals and advanced features when you get chance.

MattSturgeon commented on plugins/utils/ccc.nix:
```diff
@@ -0,0 +1,50 @@
+      lsp = false;
+    };
+  };
+  extraConfig = cfg: {opts.termguicolors = lib.mkDefault true;};
```
nit: I'd add a blank line above for consistency

MattSturgeon commented on plugins/utils/ccc.nix:
```diff
@@ -0,0 +1,50 @@
+    '';
+
+    lsp = helpers.defaultNullOpts.mkBool true ''
+      Whether to enable nvim-lsp support. The color information is updated in the background and the result is used by `|:CccPick|` and highlighter.
```
Same comments as below

traxys commented:
Note that we can squash the commits when merging too

MattSturgeon commented on plugins/utils/ccc.nix:
```diff
@@ -0,0 +1,50 @@
+      '';
+
+      lsp = helpers.defaultNullOpts.mkBool true ''
+        If true, highlight using nvim-lsp. If LS with the color provider is not attached to a buffer, it falls back to highlight with pickers. See also `|ccc-option-lsp|`.
```
- `nvim-lsp` -> `LSP` 
- `If LS` -> `If a language server`
- Description may be long enough to benefit from some line breaks?
- Does it make sense to include vimdoc links here? Or should it just be "\`:CccPick\`"

MattSturgeon commented on plugins/utils/ccc.nix:
```diff
@@ -0,0 +1,50 @@
+
+    highlighter = {
+      auto_enable = helpers.defaultNullOpts.mkBool false ''
+        Whether to enable automatically on BufEnter.
```
Nit: Wrap "`BufEnter`" in backtiicks

MattSturgeon commented on plugins/utils/ccc.nix:
```diff
@@ -0,0 +1,50 @@
+
+    highlight_mode = helpers.defaultNullOpts.mkStr "bg" ''
+      Option to highlight text foreground or background. It is used to output_line and highlighter.
+      Options: `"fg" | "bg" | "foreground" | "background"`
```
This could be an `enum` if those are the only valid values.

Also same comments regarding vim doc links (see below)

MattSturgeon commented on plugins/utils/ccc.nix:
```diff
@@ -0,0 +1,50 @@
+
+  settingsOptions = {
+    default_color = helpers.defaultNullOpts.mkStr "#000000" ''
+      The default color used when a color cannot be picked. It must be HEX format.
```
Should there be a warning/assertion that this is a valid hex string?

traxys replied:
I don't think we ever check the validity of strings (apart from enumerations)

MattSturgeon reviewed (changes requested):
Looking good. I've left some comments on the options and nit-picked their descriptions a little.

janurskremer commented:
I changed the option to enum and addressed some of the comments on formatting. I wasn't sure how to format the text, went with 80 char max line width were it seemed appropriate. Is there a guideline for such styling?
